### PR TITLE
add RespawnFlags to PlayerRespawnEvent

### LINKED ISSUE #5494: Add PlayerRespawnEvent#isEndPortalSpawn or similar
### Is your feature request related to a problem?

For the longest time, PlayerRespawnEvent is fired from when entering an End Portal after killing the Ender Dragon and currently is zero API methods of checking for this so it's impossible to only run code for actual normal respawns and not include the End Portal.

### Describe the solution you'd like.

Have some API to tell if the respawn was a normal respawn or from the end portal such as adding "isEndPortalSpawn"

### Describe alternatives you've considered.

Marking people when they enter the portal somehow but seems very hacky.

### Agreements

I have searched for and ensured there isn't already an open issue regarding this.
I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### PR DESCRIPTION
**Pull request by Machine-Maker**
Closes #5494

### CODE REVIEW

**Comment by Camotoy** [comment]
For the sake of future-proofing, it seems like it would be a better idea to have an enum for the type of respawn? That way there doesn't have to be a boolean for each type of respawn that might be an option in future Minecraft versions.

**Comment by Machine-Maker** [comment]
Its not really a *type* of respawn. This event shouldn't even be called for that I dont think, but changing that now would break compatibility.

**Comment by Camotoy** [comment]
I would propose that the enum be used for this and all future respawn options - I understand that backwards compatibility has to be maintained.

**Comment by Proximyst** [comment]
I like Camotoy's suggestion, perhaps call it `RespawnCause` or `RespawnLocationType`? Mojang might very well add more possible "location sources" for where to respawn later on, and an enum makes it easier to switch on behaviour.

**Comment by Machine-Maker** [comment]
what if more than one is true? Like its a bed spawn **AND** its from the end portal?

**Comment by kashike** [comment]
`Set<RespawnFlag>`

**Comment by Machine-Maker** [comment]
Ok, done

File: Spigot-API-Patches/0288-add-RespawnFlags-to-PlayerRespawnEvent.patch [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+         this(respawnPlayer, respawnLocation, isBedSpawn, false);
+     }
+ 
++    @Deprecated // Paper
+     public PlayerRespawnEvent(@NotNull final Player respawnPlayer, @NotNull final Location respawnLocation, final boolean isBedSpawn, final boolean isAnchorSpawn) {
++        // Paper start
++        this(respawnPlayer, respawnLocation, isBedSpawn, isAnchorSpawn, new org.bukkit.event.player.PlayerRespawnEvent.RespawnFlag[]{});
++    }
++
++    public PlayerRespawnEvent(@NotNull final Player respawnPlayer, @NotNull final Location respawnLocation, final boolean isBedSpawn, final boolean isAnchorSpawn, @NotNull final RespawnFlag...respawnFlags) {
```
Comment (kashike) [inline Spigot-API-Patches/0288-add-RespawnFlags-to-PlayerRespawnEvent.patch]: `Set<RespawnFlag>` instead - using varargs in an event constructor is a bad idea.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0288-add-RespawnFlags-to-PlayerRespawnEvent.patch (hunk 1 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0288-add-RespawnFlags-to-PlayerRespawnEvent.patch]: Done

**Review by kashike — APPROVED** [review]

File: Spigot-API-Patches/0291-add-RespawnFlags-to-PlayerRespawnEvent.patch [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ 
+     /**
+@@ -77,4 +90,31 @@ public class PlayerRespawnEvent extends PlayerEvent {
+     public static HandlerList getHandlerList() {
+         return handlers;
+     }
++
++    // Paper start
++    /**
++     * Gets whether the respawn event was caused by the end portal.
```
Comment (lynxplay) [inline Spigot-API-Patches/0291-add-RespawnFlags-to-PlayerRespawnEvent.patch]: The javadocs should probably be updated to reflect on the `RespawnFlag` concept.
As the collection seems mutable (tho modifications are not respected) this should also be stated in the documentation.

**Review by lynxplay — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0291-add-RespawnFlags-to-PlayerRespawnEvent.patch (hunk 2 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0291-add-RespawnFlags-to-PlayerRespawnEvent.patch]: I guess the collection doesn’t have to be mutable. Changing them doesn’t do anything, at the moment it’s only an informative set of flags.

Comment (lynxplay) [reply Spigot-API-Patches/0291-add-RespawnFlags-to-PlayerRespawnEvent.patch]: In my opinion it does not really matter whether or not the collection is immutable, just wanted to have the documentation state it 😂

Comment (Machine-Maker) [reply Spigot-API-Patches/0291-add-RespawnFlags-to-PlayerRespawnEvent.patch]: so ya, did that. Also changed it to an immutable set.

**Review by lynxplay — APPROVED** [review]

**Review by kennytv — APPROVED** [review]

**Pull request merged**